TimeOfFlight/src/TOFEstimators.cc

```diff
@@ -33,442 +43,578 @@
+    ts = track->getTrackState(TrackState::AtCalorimeter);
+    double phiCalo = ts->getPhi();
+    double omegaCalo = ts->getOmega();
+    double tanLCalo = ts->getTanLambda();
+
+    return abs((phiIP - phiCalo)/omegaCalo)*sqrt(1. + tanLCalo*tanLCalo);
+}
+
+
+double TOFEstimators::getTOFClosest(const Track* track, const Cluster* cluster){
+    const TrackState* ts = track->getTrackState(TrackState::AtCalorimeter);
+    Vector3D trackAtCaloPos = ts->getReferencePoint();
 
-  streamlog_out(MESSAGE) << "TOFEstimators::end()  " << name() 
-			 << " processed " << _nEvt << " events in " << _nRun << " runs "
-			 << std::endl ;
+    double dToImpactMin = 99999.;
+    double hitTime = 0.;
+    for (auto&& hit : cluster->getCalorimeterHits() ){
+        CHT hitType( hit->getType() );
+        bool isEcal = (hitType.caloID() == CHT::ecal);
+        if (!isEcal) continue;
 
+        Vector3D hitPos = hit->getPosition() ;
+        double dToImpact = (hitPos - trackAtCaloPos).r();
+
+        if(dToImpact < dToImpactMin){
+            dToImpactMin = dToImpact;
+            hitTime = hit->getTime();
+        }
+    }
+    hitTime += _smearing(_generator);
+    return hitTime - dToImpactMin / SPEED_OF_LIGHT;
 }
 
 
-//*******************************************************************************************************
+double TOFEstimators::getTOFFastest(const Track* track, const Cluster* cluster){
+    const TrackState* ts = track->getTrackState(TrackState::AtCalorimeter);
+    Vector3D trackAtCaloPos = ts->getReferencePoint();
+
+    double dToImpactMin = 0.;
+    double hitTimeMin = 99999.;
+    for (auto&& hit : cluster->getCalorimeterHits() ){
+        CHT hitType( hit->getType() );
+        bool isEcal = (hitType.caloID() == CHT::ecal);
+        if (!isEcal) continue;
+
+        Vector3D hitPos = hit->getPosition();
+        double dToImpact = (hitPos - trackAtCaloPos).r();
+        double hitTime = hit->getTime() + _smearing(_generator);
+        if(hitTime < hitTimeMin){
+            dToImpactMin = dToImpact;
+            hitTimeMin = hitTime;
+        }
+    }
+    return hitTimeMin - dToImpactMin / SPEED_OF_LIGHT;
+}
+
 
+double TOFEstimators::getTOFCylFit(const Track* track, const Cluster* cluster){
+    const TrackState* ts = track->getTrackState(TrackState::AtCalorimeter);
+    Vector3D trackAtCaloPos = ts->getReferencePoint();
+    Vector3D trackAtCaloMom = getMomAtCalo(track);
+
+    vector <double> x, xErr, y, yErr;
+
+    for (auto&& hit : cluster->getCalorimeterHits() ){
+        CHT hitType( hit->getType() );
+        bool isEcal = (hitType.caloID() == CHT::ecal);
+        int layer = hitType.layer();
+        if (!isEcal || layer >= _maxLayerNum) continue;
+
+        Vector3D hitPos = hit->getPosition();
+        double dToLine = (hitPos - trackAtCaloPos).cross(trackAtCaloMom.unit()).r();
+        // take only hits from 5 mm cylinders. 5 mm value was ibtain optimizing on photons
+        if (dToLine > 5.) continue;
```

Comment:
Should be moved to a processor parameter maybe (?). Not nice to have something hardcoded